### PR TITLE
First step at adding exceptions

### PR DESCRIPTION
**Pull request by eellison**
This is a first step towards adding exceptions. We need minimal support in order to begin converting the torch library to weak script mode (which is the main goal here).  

Some limitations (that are documented in the tests & compiler): 
1. Cannot assign exceptions to variables
2. Any name after raise is being treated as a valid Exception 
3. No control flow analysis yet. Below a will be undefined:

if True:
     a = 1
else:
     raise Exception("Hi")
return a

### CODE REVIEW

File: torch/csrc/jit/register_prim_ops.cpp [hunk 1]
```diff
@@ -233,18 +245,27 @@ RegisterOperators reg({
...
-              std::cout << i;
-            }
+            printListToStream(std::cout, " ", last(stack, num_inputs));
             drop(stack, num_inputs);
             std::cout << std::endl;
             return 0;
           };
         }),
+    Operator(
+        prim::RaiseException,
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: I think this should just raise an exception and take a single string argument. String formatting should be done externally to this op.

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -1308,6 +1311,30 @@ struct to_ir {
...
+  // raise AnyName(3)
+  //
+  // NYI: add exception logic to control-flow nodes
+  // if True:
+  //   a = 1
+  // else
+  //   raise Exception("Hi")
+  // print(a)
+  void emitRaise(const Raise& stmt) {
+    auto apply = Apply(stmt.expr());
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: There is no guarantee that `stmt.expr()` is an apply node. I can write `raise 3 + 4`

File: torch/jit/frontend.py [hunk 3]
```diff
@@ -266,6 +266,15 @@ def build_Return(ctx, stmt):
         values = (stmt.value,) if not isinstance(stmt.value, ast.Tuple) else stmt.value.elts
         return Return(r, [build_expr(ctx, val) for val in values if val is not None])
 
+    @staticmethod
+    def build_Raise(ctx, stmt):
+        r = ctx.make_range(stmt.lineno, stmt.col_offset, stmt.col_offset + len("raise"))
+        if isinstance(stmt.exc, ast.Name):
```
Comment (zdevito) [inline torch/jit/frontend.py]: It is weird to check for ast.Name but not to disallow all the other expressions that we do not handle (`1`, `a + b`, etc.). For now, we allow anything at this stage, and then just ignore the expression.

File: torch/csrc/jit/script/compiler.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: In python raise has to take in an exception. raise 3 + 4 doesn't work

Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: Good point, updated

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -1314,6 +1317,34 @@ struct to_ir {
...
+  // else
+  //   raise Exception("Hi")
+  // print(a)
+  void emitRaise(const Raise& stmt) {
+    if (stmt.expr().kind() != TK_APPLY) {
+      throw ErrorReport(stmt) << "Invalid exception. Exceptions cannot "
+        "be constructed outside of a raise statement";
+    }
+    auto apply = Apply(stmt.expr());
+    auto exception_name = Var(apply.callee()).name().name();
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Now you are allowing anything that must be a call node in the parser. But you are not forcing the thing being called to be a variable. This now has the same problem as above. Just ignore the entire exception statement.

File: test/test_jit.py [hunk 5]
```diff
@@ -7355,6 +7357,54 @@ def forward(self, x, y, c):
...
+    def test_exceptions(self):
+        cu = torch.jit.CompilationUnit('''
+            def foo(cond):
+                if bool(cond):
+                    raise ValueError(3)
+                return 1
+        ''')
+
+        cu.foo(torch.tensor(0))
+        with self.assertRaisesRegex(RuntimeError, "Exception"):
```
Comment (apaszke) [inline test/test_jit.py]: Can we at least throw a `JITException`? That will be a clearer change than defaulting to `RuntimeError` for anything that was written inside.

File: torch/csrc/jit/passes/constant_propagation.cpp [hunk 6]
```diff
@@ -16,17 +16,10 @@ std::unordered_set<Symbol> skip_list = {
   prim::If,
   prim::Loop, //TODO: handle Loop
   prim::Print,
+  prim::RaiseException,
   prim::PythonOp, //may have side effects
   prim::LoadWorld,
   prim::StoreWorld,
-  //all the rand functions from native_functions.yaml
```
Comment (driazati) [inline torch/csrc/jit/passes/constant_propagation.cpp]: Why are these deleted in this PR?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 7]
```diff
@@ -234,17 +235,26 @@ RegisterOperators reg({
...
+        [](Node* node) -> Operation {
+          return [](Stack& stack) {
+            throw JITException(pop(stack).toString()->string());
```
Comment (driazati) [inline torch/csrc/jit/register_prim_ops.cpp]: `toStringRef`

File: torch/jit/frontend.py [hunk 8]
```diff
@@ -266,6 +266,12 @@ def build_Return(ctx, stmt):
         values = (stmt.value,) if not isinstance(stmt.value, ast.Tuple) else stmt.value.elts
         return Return(r, [build_expr(ctx, val) for val in values if val is not None])
 
+    @staticmethod
+    def build_Raise(ctx, stmt):
+        r = ctx.make_range(stmt.lineno, stmt.col_offset, stmt.col_offset + len("raise"))
+        expr = build_expr(ctx, stmt.exc)
```
Comment (driazati) [inline torch/jit/frontend.py]: Python 2 doesn't have `exc` on `Raise` nodes ([docs](https://docs.python.org/2/library/ast.html#abstract-grammar)) so this throws an error. Probably best to figure how to do the same thing in Python 2 and put the functionality in `_six.py`

**Comment by driazati** [comment]
This looks good for a first pass at Exceptions, but needs fixes for the failing tests / Python 2 stuff. Also needs a follow up PR for exception types / messages once we get the standard library done.

File: torch/csrc/jit/passes/constant_propagation.cpp (hunk 6 above)
Comment (eellison) [reply torch/csrc/jit/passes/constant_propagation.cpp]: I switched it to use the more comprehensive set of random ops

File: torch/csrc/jit/script/compiler.cpp [hunk 9]
```diff
@@ -1295,6 +1298,27 @@ struct to_ir {
...
+  //
+  // NYI: add exception logic to control-flow nodes
+  // if True:
+  //   a = 1
+  // else
+  //   raise Exception("Hi")
+  // print(a)
+  void emitRaise(const Raise& stmt) {
+    const std::string exception = "Exception";
+    auto string_input = insertConstant(*graph, exception, stmt.range());
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: ehe source location gets added to the exception message when it is thrown

**Review by driazati — APPROVED** [review]

File: torch/csrc/jit/interpreter.cpp [hunk 10]
```diff
@@ -664,10 +665,14 @@ struct InterpreterStateImpl {
...
           pc = new_pc;
         } catch(std::exception & e) {
-          if(!instructions[pc].debug_location)
-            throw; // rethrow original exception
-          // throw a new exception with enhanced debugging information
-          instructions[pc].debug_location->wrapAndRethrowException(e, "operation failed in interpreter");
+          if (!instructions[pc].debug_location) {
+            throw;
+          }
+          if (JITException* jit_e = dynamic_cast<JITException *>(&e)) {
```
Comment (zdevito) [inline torch/csrc/jit/interpreter.cpp]: Both branches of this if do the same thing. I don't think it is needed.

File: torch/csrc/jit/source_location.h [hunk 11]
```diff
@@ -14,13 +16,22 @@ namespace torch { namespace jit {
...
-    std::stringstream msg;
+
+  void wrapException(std::ostream& msg, const std::exception & e, const std::string & additional = "") {
     msg << "\n" << e.what() << ":\n";
     if(!additional.empty()) {
       msg << additional << ":\n";
     }
     highlight(msg);
+  }
+  void wrapAndRethrowException(const JITException & e, const std::string & additional = "") {
```
Comment (zdevito) [inline torch/csrc/jit/source_location.h]: same here: two copies of the same exception.

**Review by zdevito — APPROVED** [review]
This looks good, but the duplicate code in `wrapAndRethrowException` logic should be removed.

File: torch/csrc/jit/source_location.h (hunk 11 above)
Comment (eellison) [reply torch/csrc/jit/source_location.h]: Since the function is re-throwing a new exception - JITException/std::runtime_error - don't you need two separate functions to differentiate the newly instantiated exception ?

Comment (zdevito) [reply torch/csrc/jit/source_location.h]: ok, I see what you are doing here, but it is a confusing organization: It overloads an operator based on a static (JITException vs exception) subtype. And the requires a static check and cast to dispatch to that subtype. It would be better if the only function in SourceLocation were `void wrapException(std::ostream& msg, const std::exception & e, const std::string & additional = "")` and then the remaining logic could go in interpreter.cpp

File: torch/csrc/jit/interpreter.cpp [hunk 12]
```diff
@@ -668,11 +668,16 @@ struct InterpreterStateImpl {
           if (!instructions[pc].debug_location) {
             throw;
           }
-          if (JITException* jit_e = dynamic_cast<JITException *>(&e)) {
-            instructions[pc].debug_location->wrapAndRethrowException(*jit_e, "operation failed in interpreter");
-          } else {
-            instructions[pc].debug_location->wrapAndRethrowException(e, "operation failed in interpreter");
+          // throw correct subtype
```
Comment (zdevito) [inline torch/csrc/jit/interpreter.cpp]: Can we make this just: 
```
std::string msg = instructions[pc].debug_location->wrapErrorMessage(e.what(), "operation failed in interpreter);
if(dynamic_cast<JITException*>(&e) {
  throw JITException(msg);
} else {
  throw std::runtime_error(msg);
}
```
It is much simpler, and doesn't have  bunch of rethrow logic that occurs across different files.
